text/0000-conditional-compilation-checking.md

```diff
@@ -0,0 +1,486 @@
+
+### Examples
+
+```bash
+rustc --cfg 'valid(cats, dogs, foo, bar)' ...
+```
+
+The `--cfg 'valid(...)'` option may be repeated. If it is repeated, then the condition names of
+all of the options are merged into a single set. For example:
+
+```bash
+rustc --cfg 'valid(cats)' --cfg 'valid(dogs)' ...
+```
+
+### Well-known condition names
+
+`rustc` defines a set of well-known conditions, such as `test`, `target_os`, etc. These conditions
+are always valid; it is not necessary to enable checking for these conditions. If these conditions 
+are specified in a `--cfg valid(...)` option, they will be ignored. This set of well-known names is
+a part of the stable interface of the compiler. New well-known conditions  may be added in the
+future, because adding a new name cannot break existing code. However, a name may not be removed
+from the set of well-known names, because doing would be a breaking change.
+
+These are the well-known conditions:
+
+* `feature`
+* `linux`
+* `test`
+* `target_os`
+* `target_arch`
+* `windows`
+* TODO: finish enumerating this list
+
+## Checking key-value conditions
+
+For conditions that define a list of values, such as `feature`, we want to verify that any
+`#[cfg(feature = "v")]` test uses a valid value `v`. We want to detect this kind of bug:
+
+```rust
+if cfg!(feature = "awwwwsome") {    // should have been "awesome"
+    ...
+}
+```
+
+This kind of bug could be due to a typo or a bad PR merge. It could also occur because a feature
+was removed from a `Cargo.toml` file, but source code still contains references to it.
+
+To catch these errors, we give `rustc` the set of valid values for a given condition name.
+To do so, we extend the syntax of the `--cfg` option. For a given condition name `c` (such as
+`feature`), we specify the set of valid values as:
+
+```bash
+rustc --cfg 'valid_values(c, "value1", "value2", ... "valueN")'
+```
+
+### Examples
+
+```bash
+rustc --cfg 'valid_values(feature, "derive", "parsing", "printing", "proc-macro")'
+
+# or equivalently:
+rustc --cfg 'valid_values(feature, "derive")' \
+    'valid_values(feature, "parsing")' \
+    'valid_values(feature, "printing")' \
+    'valid_values(feature, "proc-macro")'
+
+rustc --cfg 'valid_values(target_os, "linux", "macos", "ios", "windows", "android")'
+
+# specifying values for different names requires more than one --cfg option
+rustc --cfg 'valid_values(foo, "red", "green")' --cfg 'valid_values(bar, "up", "down")'
+```
+
+## Checking is opt-in (disabled by default)
+
+The default behavior of `rustc` is that conditional compilation names and values are not checked.
+This maintains compatibility with existing versions of Cargo and other build systems that might
+invoke `rustc` directly. All of the information for checking conditional compilation uses new
+syntactic forms of the existing `--cfg` option.
+
+Checking condition names is independent of checking condition values, for those conditions that
+use value lists. Examples:

```

Comment:
I think this should be a lint that's warn-by-default. That lets people disable it while still making it discoverable.

Reply:
That would mean that Cargo would immediately start reporting warnings in situations where it didn't before. That would surprise people, right?  Is that typically how new lints are added?

Reply:
Yes, typically new lints are warn-by-default unless there's a strong reason for them not to be. Warnings are not part of Rust's stability policy. https://github.com/rust-lang/rust/issues/43466#issuecomment-643899661

Reply:
Sounds good.  I'll update the doc.